Assets loading with themes and skins supported

Yahasana opened the pull request:
Assets loading with themes and skins supported. assume that different apps (or plugins) are independent with each others but share the same basic modules and public resources

A) asset cases: there are stylesheet files in public `css` share by all apps 
and app specific stylesheet locate in its directory. e.g.
 -  public/css/style.css
 -  public/min/css/style.min.css
 -  app/account/skin/default/css/account.css
 -  public/min/css/account.min.css

B) loading source and target
   1.  load the min file
        - a) load from public:  /css/style.css ===> public/min/css/style.min.css
              minify `public/css/style.css` to `public/min/css/style.min.css`
        - b) load from app dir: css/account.css ===> public/min/css/account.min.css
              minify `account/default/css/account.css` to `public/min/css/account.min.css`
   2.  load the raw file
        - a) load from public: /css/style.css ===> public/css/style.css
        - b) load from app dir: default/css/account.css ===> public/default/css/account.css
            - i) if `public/default/css/account.css` not exists, 
                copy `account/default/css/account.css` to `public/default/css/`
            - ii) else update `account/default/css/account.css` to `public/default/css/`

C) implement

1) if `uri` is start with `/`, it'll try to load asset from document root
```php

$mySkin = 'Default';

// Register specific services for the module
public function registerServices(Di $di)
{
    $di->assets->setOptions([
            'source' => __DIR__ . '/Skin/' . $mySkin . '/',
            'target' => 'min/',
            'minify' => 3
        ]);
}

// css used by specific module
// load from module dir/Skin/Default/css/account.css
$this->assets->addCss(['css/account.css']); 

// css shared by all modules
// load from document_root/Skin/Default/css/style.css
$this->assets->addCss(['/Skin/' . $mySkin . '/css/style.css']); 
```
2) always store the min files in the document_root/target dir

**ONE MORE THING**

`type` parameter change from `text/css` or `text/javascript` to `css` or `js`
```php
$assets->addCss(['css/style.css', 'type' => 'css'])
->addCss(['js/script.js', 'type' => 'js'])
```
but not
```php
$assets->addCss(['css/style.css', 'type' => 'text/css'])
->addCss(['js/script.js', 'type' => 'text/javascript'])

Commit:
load assets from document root dir

1) if `uri` is start with `/`, it'll try to load asset from document root
```php
// Register specific services for the module
public function registerServices(Di $di)
{
    $di->assets->setOptions([
            'source' => __DIR__ . '/Skin/Default/',
            'target' => 'min/',
            'minify' => 3
        ]);
}

// load from module dir/Skin/Default/css/style.css
$this->assets->addCss(['css/style.css']); 

// load from document_root/Skin/Default/css/style.css
$this->assets->addCss(['/Skin/Default/css/style.css']); 
```
2) always store the min files in the document_root/target dir

Commit:
starts_with

Commit:
target is start from absolute path

place min file to document root

Commit:
fixed missing / for source

Yahasana commented:
@mruz please review the new features i just add to comment. i'll trying to implement it soon if you agree with that. and after that i think **Ice** is ready for new release and good for deploying to production

mruz commented on ice/assets.zep:
```diff
@@ -258,9 +258,28 @@ class Assets
-            destination = source . uriMin,
             exist = false;
 
+        // there is no source dir, try to load from document root
```
Will it work if we just add default values?
```zep

         let source = this->getOption("source", _SERVER["DOCUMENT_ROOT"]),
             target = this->getOption("target", "min" . DIRECTORY_SEPARATOR),
```

Commit:
Auto version the uri

get/set cache the prepared resources

Commit:
suspend error is evil

Commit:
one more let

Commit:
Syntax error

Commit:
undefnined variable 'type'

Commit:
md5file

Commit:
DIRECTORY_SEPARATOR

Commit:
make target dir equal to 'min'

Commit:
replace \ to /

Commit:
if no target, place versioned file to the same dir

Yahasana commented:
@mruz review plz

Commit:
bugs fixed

`type` parameter change from `text/css` or `text/javascript` to `css` or `js`
```php
$assets->addCss(['css/style.css', 'type' => 'css'])
->addCss(['js/script.js', 'type' => 'js'])
```
but not
```php
$assets->addCss(['css/style.css', 'type' => 'text/css'])
->addCss(['js/script.js', 'type' => 'text/javascript'])
